DOCSP-45004: Run Command (#320)
mongodb/docs-csharp: source/fundamentals/databases-collections/run-command.txt

.. important:: Prefer Driver Methods to Database Commands

   The driver provides wrapper methods for many database commands.
   If possible, we recommend using these methods instead of executing
   database commands.
   
   To perform administrative tasks, use the :mongosh:`MongoDB Shell </>`
   instead of the {+driver-short+}. The shell provides helper methods
   that might not be available in the driver.
   
   If there are no available helpers in the driver or the shell, you
   can use the ``db.runCommand()`` shell method or the driver's
   ``RunCommand()`` and ``RunCommandAsync()`` methods, which are described in this
   guide.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``sample_restaurants.restaurants`` collection
from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
free MongoDB Atlas cluster and load the sample datasets, see the :ref:`<csharp-quickstart>`.

Execute a Command
-----------------

To run a database command, create a ``BsonDocument`` object that specifies
the command and pass it as a parameter to the ``RunCommand()`` or ``RunCommandAsync()`` 

> method. You can specify the type returned by the ``RunCommand()`` and ``RunCommandAsync()``
> methods by specifying the type parameter. You can use the ``BsonDocument`` type to return

Reviewer: [s] Suggestion for brevity and reducing repetition:
```suggestion
method. You can specify the type returned by these methods by specifying the type parameter. You can use the ``BsonDocument`` type to return
```

   .. tab:: Synchronous
      :tabid: distinct-sync

      .. literalinclude:: /includes/fundamentals/code-examples/databases-collections/RunCommand.cs
         :language: csharp
         :dedent:
         :start-after: start-read-pref
         :end-before: end-read-pref

.. tip::

   To learn more about read preference options, see :manual:`Read
   Preference </core/read-preference/>` in the {+mdb-server+} manual.

.. _csharp-command-response:

Response
--------

The raw command response document returned by the ``RunCommand()`` method contains the
following fields:

.. list-table::
   :header-rows: 1
   :widths: 30 70

   * - Field
     - Description

   * - ``<command result>``
     - Fields specific to the database command. For example,
       the ``hello`` command returns the ``topologyVersion`` field.

   * - ``ok``
     - Indicates whether the command has succeeded (``1.0``) or failed (``0.0``). The
       driver raises a ``MongoCommandException`` if the ``ok``
       value is ``0.0``.

   * - ``$clusterTime``
     - A document that contains the signed cluster time. Cluster time is a
       logical time used for the ordering of operations. This field only

>        applies to commands run on replica sets or sharded cluster.
> 
>    * - ``operationTime``
>      - The logical time of the operation execution. This field only
>        applies to commands run on replica sets or sharded cluster.

Reviewer: [q] Are these meant to be singular on purpose? If not, would update:
```suggestion
       applies to commands run on replica sets or sharded clusters.

   * - ``operationTime``
     - The logical time of the operation execution. This field only
       applies to commands run on replica sets or sharded clusters.
```
